Mortar augmented lagrange

dschwen opened the pull request:
Rewrite penalty mortar formulation using Simo's paper, implement augmented Lagrange for normal contatct.

Refs  #24168

dschwen committed:
Add PMAL interface (#24168)

dschwen committed:
Implement normal Lagrange multiplier (#24168)

dschwen committed:
wip

dschwen committed:
virtual inheritance

dschwen committed:
use AD

dschwen committed:
update multiplier

dschwen committed:
Template problem class

dschwen committed:
benchmark

dschwen committed:
Add linear predictor for Lagrange multiplier (#24168)

dschwen committed:
Implement Simo's penalty algorithm using Eigen (#24168)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1526286/) on ece6bb1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b9c8e604ca417526aacf5779b79226cc683acf26`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1526328/) on bab5234 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b9c8e604ca417526aacf5779b79226cc683acf26`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1526437/) on db47787 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 587ef504e9926d4c3e2b25ffff65122ba09a9a63`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1526546/) on 62e0907 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24349/site)

This comment will be updated on new commits.

dschwen commented:
```
contact/test:pdass_problems.cylinder_friction_penalty
```

Is the only test that needs to be regolded.

dschwen commented:
> ```
> test:3d-mortar-contact.frictional-mortar-3d-penalty
> ```
> 
> Now fails to converge :-o

```
accumulated_slip_one             rel diff:  4.5256383e-05 ~ 4.8240713e+228 = 1.00000e+00
```

ha, looks like something is uninitialized...

dschwen committed:
Zero initialize vectors (#24168)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1526730/) on 30a6a5c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 587ef504e9926d4c3e2b25ffff65122ba09a9a63`

recuero commented on modules/contact/include/userobjects/PenaltyFrictionUserObject.h:
```diff
@@ -52,33 +56,19 @@ class PenaltyFrictionUserObject : public WeightedVelocitiesUserObject
-
-  /// Map from degree of freedom to old accumulated slip
-  std::unordered_map<const DofObject *, std::array<ADReal, 2>> _dof_to_old_accumulated_slip;
+  /// Map from degree of freedom to current and old accumulated slip
```
It'd be nice to indicate in the comment that slip is only used for output purposes.

recuero commented on modules/contact/src/problems/AugmentedLagrangianContactProblem.C:
```diff
@@ -145,6 +165,24 @@ AugmentedLagrangianContactProblem::checkNonlinearConvergence(std::string & msg,
 
+      // next loop over penalty mortar user objects
+      std::vector<MortarUserObject *> mortar_uos;
+      std::list<PenaltyMortarAugmentedLagrangeInterface *> pmuos;
```
Does this assume that all penalty friction mortar user objects will be AL?

dschwen replied:
No, only the ones that inherit form the interface

recuero commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -78,15 +114,82 @@ PenaltyWeightedGapUserObject::reinit()
+  for (const auto & [dof_object, gap] : _dof_to_weighted_gap)
+    if (physicalGap(gap) < -_penetration_tolerance ||
+        (physicalGap(gap) > _penetration_tolerance && _dof_to_lagrange_multiplier[dof_object] > 0))
+      return false;
```
It'd be nice to have a machinery to report the least converged node at the end of the time step or maybe also at the end of the nonlinear iteration. Just a thought. Not to be done in this PR.

recuero commented on modules/contact/test/tests/sliding_block/in_and_out/frictional_02_penalty_out.e:
Why do we need to update this node-face results? Syntax changes?

dschwen replied:
That's not in the current PR. Not sure why that file was ever touched :-/

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1531041/) on 8a9049b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8da381</th>
      <th colspan=4>#24349 8a9049</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.21%</td>
      <td>85.21%</td>
      <td>-0.00%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87795</td>
      <td>87798</td>
      <td>+3</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15242</td>
      <td>15245</td>
      <td>+3</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24349/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24349/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>8da381</th>
      <th colspan=4>#24349 8a9049</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.99%</td>
      <td>89.24%</td>
      <td>-0.76%</td>
      <td>76.42%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3543</td>
      <td>3507</td>
      <td>-36</td>
      <td>94</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>394</td>
      <td>423</td>
      <td>+29</td>
      <td>29</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24349/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24349/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24349/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 66.67% is less than the suggested 90.0%
- `contact` new line coverage rate 76.42% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented:
Ugh, regolding the cylinder test I noticed that the result for a serial run are different (of course we're not testing that)

dschwen committed:
Augmented frictional LM (#24168)

moosebuild commented:
Job [Newer gcc](https://civet.inl.gov/job/1539755/) on 1e066b0 : invalidated by @dschwen

dschwen committed:
Regold cylinder test (#24168)

moosebuild commented:
Job [Python 3.8](https://civet.inl.gov/job/1540293/) on 7c4292f : invalidated by @dschwen

fails repeatedly

dschwen committed:
Add and refactor output (#24168)

dschwen committed:
Rework tests (#24168)

dschwen committed:
refine contact area

dschwen committed:
wip

dschwen committed:
more wip

dschwen committed:
code fixes for Uszawa - still not right

dschwen committed:
Add input and jupyter notebook

dschwen committed:
aux variable for active set

recuero committed:
Penalty implementation passes penalty tests

Small changes to code, tests, and gold files to make the penalty
implementation that enables augmented Lagrange work with existing
mortar penalty tests.

Small differences in the penalty friction implementation appear to
exist for stick behavior, but it is on the order of 1.0e-3. This
might be produced by the fact that for AL we cache the force instead
of the accumulated slip and some ensuing roundoff error.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1593286/) on 46e83d1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 23ac2e6b1e6eca40969c8f4bacad2a8d5112b4a1`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1593286/) on 46e83d1 wanted to post the following:

The following commit(s) contain fixups:

- 88d15359d1: fixup! Rework tests (#24168)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1593374/) on 4954eb4 wanted to post the following:

The following commit(s) contain fixups:

- 88d15359d1: fixup! Rework tests (#24168)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1593466/) on cb745f6 wanted to post the following:

The following commit(s) contain fixups:

- 88d15359d1: fixup! Rework tests (#24168)

recuero committed:
Results are smooth with AL and initially verified against penalty.

Removed active set from the normal force. Active set information can be
retrieved from final contact pressure value e.g. (if > 0, active).

I noticed there is an artifact at the edge of the contact patch if the
penalty coefficient is very high. It looks as if there is a kink even though
the gap is identically satisfied. It may have to do with the mortar segment
mesh. We should keep an eye on this for other models.

I didn't follow exactly Uzawa's algorithm, so improvements are obviously possible.

We should:

 - Have more tests, bouncing block test and verification against LM formulation.
 - Take another look at Uzawa's algorithm. It is advertised as being superior in
   a system Newton loop, which is what we are doing.
 - Add convergence criterion for friction (the one now is fake/trivial) !!
 - Tests get stuck in parallel (revisit this).

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1594104/) on 3d4eb9c wanted to post the following:

The following commit(s) contain fixups:

- 88d15359d1: fixup! Rework tests (#24168)

recuero committed:
Add test for AL that checks separation

dschwen committed:
Fix parallel bug

recuero committed:
Local commit to be revised

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1605565/) on 5f917c2 : invalidated by @dschwen

recuero committed:
Use dual bases for AL contact

recuero committed:
Modify AL frictional algorithm.

- Account for sign of normal_lm in the update formula of the LMs
- Make sure frition is only applied to the active set normal_lm_node < 0
- Allow some online adaptivity of the local penalty friction coefficient (to be improved)

We could fix the frictional state of each node for each outer loop. This remains to
be done.

recuero committed:
Add AL to action with dual bases with test

recuero committed:
For penalty, use "normalized" normal penalty value.

That is, in addition to get the physical gap, we divide by a value given
by normalize_c, which is a proxy for the physical area that surrounds that
node. This is done at the request of users that are familiarized with the
normalized_penalty option of node-face. The values would not be exactly the
same, but it should allow for using the same penalty value in both formulations.

For a mechanics problem, the normal penalty coefficient can be selected on the order
of the modulus of elasticity of the materials. In the case of AL, it should be
selected as the softest of the materials in contact.

Also:

Adjust tolerances and use absolute slip.

Tolerances are adjusted because there are nonzero quantities that
range from 1.0e-4 to 1.0e5, such as pressure, so we add a abs zero to
those tests.

recuero committed:
Add AL example with hyper algebraic multiplier as preconditioner

recuero committed:
Add documentation on mortar penalty/AL.

Add missing exodus file.

The adaptivity parallel test is regolded. That's necessary because, now,
the action defaults to physical gap/normalized penalty, so the diffs
are due to different penalty coefficients in practice.

recuero committed:
Avoid sweep failure of an 'error' output test

recuero committed:
Small changes to AL frictional logic:

 - Tolerance checks
 - Adapted penalty for slip distance
 - Add a test that _converges_ on the slip distance

recuero committed:
Add AL test to combined module's thermomechanics vertical setup

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1655652/) on b79958a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24349/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d033983fbc0a22e6cbea27126bdc0421ba44e1c8`

moosebuild commented:
Job [Conda MOOSE (ARM Mac)](https://civet.inl.gov/job/1655708/) on d78524c : invalidated by @recuero

dschwen committed:
Fix min AD sizes in tests (#24168)

dschwen committed:
Compute normal pressure once (#24168)

dschwen committed:
Parallel consistent messages (#24349)

dschwen committed:
Improve coverage (#24349)

dschwen commented:
Ok, @bwspenc , coverage is up. The one failing test doesn't exist according to @loganharbour , so just ignore that.

loganharbour commented:
> The one failing test doesn't exist 

😢

moosebuild commented:
Job [Conda MOOSE (ARM Mac)](https://civet.inl.gov/job/1671980/) on c874ff3 : invalidated by @recuero

bwspenc commented:
You should remove "WIP" from the PR description now

recuero commented:
It has been removed.

bwspenc commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -68,25 +143,152 @@ PenaltyWeightedGapUserObject::getNormalContactPressure(const Node * const node)
+bool
+PenaltyWeightedGapUserObject::isAugmentedLagrangianConverged()
+{
+  Real positive_gap = 0.0;
```
I think this would be more clear if these had 'max' or 'extreme' or something in their names to better convey what they are.

bwspenc commented on modules/contact/test/tests/verification/patch_tests/automatic_patch_update/tests:
```diff
@@ -28,6 +28,8 @@
                  "again."
     prereq = sliding_blocks
     min_parallel = 2
+    # max parallel set to avoid not running into error for more procs
```
I'm not really sure what this means. It errors with more procs? Is that a problem? What is the error?

recuero replied:
It stops erroring with more processors. This issue only appears for some decompositions. Since we are checking an error, let's only choose the domain decompositions that actually cause such an error. So it's indeed "avoiding not running into errors for more processors."

bwspenc replied:
Something like "The error being tested does not occur for larger numbers of processors" would be more clear.

recuero replied:
Yeah, we can make that change to make the message more clear. However, "The error being tested does not occur for larger numbers of processors" isn't accurate. It may occur for a larger number of processors at some point.

recuero commented on modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d-rz/finite_action_al.i:
```diff
@@ -0,0 +1,307 @@
+
+[MortarGapHeatTransfer]
+  [mortar_heat_transfer]
+    temperature = temp
```
No, gap heat transfer is independent of the ALM for mechanical contact.

recuero commented on modules/contact/doc/content/source/userobjects/PenaltyFrictionUserObject.md:
```diff
@@ -9,17 +9,55 @@ frictional pressure(s) saturates at the Coulomb limit. Note that a high enough
 (see [ComputeFrictionalForceLMMechanicalContact](/ComputeFrictionalForceLMMechanicalContact.md)).
 This object is set up automatically when using the contact action [ContactAction](/ContactAction.md).
 
+This object is built from within the contact action when the formulation `MORTAR_PENALTY` and
```
We can put backticks or have a hyperlink. But capitalization of 'contact' when referring to the contact action shouldn't be done according to MOOSE PR review feedback.

dschwen replied:
We've linked to the contact action page above already.

recuero commented on modules/contact/doc/content/source/userobjects/PenaltyFrictionUserObject.md:
```diff
@@ -9,17 +9,55 @@ frictional pressure(s) saturates at the Coulomb limit. Note that a high enough
+taking the necessary nonlinear iterations and updates "fixed" Lagrange multipliers in an outer loop. This
+process repeats until the contact-related tolerances are met. The "fixed" Lagrange multipliers represent
+accumulated normal and tangential tractions over the AL iterations (see [!citep](wriggers2006computational)).
+Usage of AL with this mortar constraint allows for, simultaneously 1. Having a consistent contact formulation,
```
It just means we retrieve the mortar enforcement with its formulation and numerical features. Will elaborate.

recuero commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -101,10 +101,18 @@ ContactAction::validParams()
+      1.0,
+      "penalty_multiplier_friction > 0",
+      "The penalty growth factor between augmented Lagrange "
+      "iterations for penalizing relative slip distance if the node is under stick conditions.");
```
> When I talked with Antonio about this, he told me that the AL algorithm adaptively increases the penalty over the course of the AL iterations.

I meant the logic is _always_ there. If the penalty multiplier is one, the factor will never change over the AL iterations. Adapting the penalty factor is key to having good convergence, so we will work on that some more.

bwspenc replied:
OK, thanks for the clarification!

recuero commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -163,174 +183,221 @@ PenaltyFrictionUserObject::getTangentialVelocity(const Node * const node,
+  PenaltyWeightedGapUserObject::reinit();
 
-  // Frictional pressure (2D)
+  // Reset frictional pressure
```
It's a pressure (contact traction) that's interpolated. We should rename the member variable.

recuero commented on modules/contact/src/userobjects/PenaltyFrictionUserObject.C:
```diff
@@ -163,174 +183,221 @@ PenaltyFrictionUserObject::getTangentialVelocity(const Node * const node,
-      _frictional_contact_force_one[qp] += test_i[qp] * frictional_nodal_pressure_one;
-      _frictional_contact_force_two[qp] += test_i[qp] * frictional_nodal_pressure_two;
+      mooseInfoRepeated(
+          "Stick tolerance fails. Slip distance for sticking node ",
```
> This is just saying that we haven't yet converged, right?

Yes. It's very easy to be too optimistic when setting these tolerances.

> We wouldn't print a message at every NL iteration suggesting that the user might want to loosen the tolerance 

The message would be printed every `AL` iteration.

recuero commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -8,22 +8,61 @@
   params.addRequiredParam<Real>("penalty", "The penalty factor");
+  params.addRangeCheckedParam<Real>(
+      "penalty_multiplier",
+      100.0,
```
The compelling reason is that the contact action already used `1`. To get good convergence, particularly with frictionless contact, having a penalty multiplier of 10-100 helps quite a bit.

recuero commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -68,25 +143,152 @@ PenaltyWeightedGapUserObject::getNormalContactPressure(const Node * const node)
+
+    // Update penalty
+    const auto previous_gap = _dof_to_previous_gap[dof_object];
+    if (std::abs(gap) > 0.25 * std::abs(previous_gap) && _lagrangian_iteration_number < 6)
```
We will go back to this. This type of magic numbers are even recommended in the textbooks.

bwspenc replied:
A short comment explaining that would be helpful.

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -787,6 +821,18 @@ ContactAction::addMortarContact()
       var_params.set<std::vector<VariableName>>("disp_y") = {displacements[1]};
       var_params.set<Real>("penalty") = getParam<Real>("penalty");
+
+      if (isParamValid("al_penetration_tolerance"))
```
No, we can add an error check if AL parameters are set and the problem class does not support AL. Right now AL is Penalty _with_ a Problem class that supports AL.

dschwen commented on modules/contact/src/postprocessors/NumAugmentedLagrangeIterations.C:
```diff
@@ -0,0 +1,36 @@
+    _augmented_lagrange_problem(
+        dynamic_cast<AugmentedLagrangianContactProblemInterface *>(&_fe_problem))
+
+{
```
I don't quite see the benefit of that error. Users might want to switch AL on and off, and having to comment out that PP seems like a hassle. `0` is a perfectly valid answer. There are _no_ AL iterations if the problem class is not an AL problem class.

dschwen commented on modules/contact/doc/content/source/userobjects/PenaltyFrictionUserObject.md:
```diff
@@ -9,17 +9,55 @@ frictional pressure(s) saturates at the Coulomb limit. Note that a high enough
 (see [ComputeFrictionalForceLMMechanicalContact](/ComputeFrictionalForceLMMechanicalContact.md)).
 This object is set up automatically when using the contact action [ContactAction](/ContactAction.md).
 
+This object is built from within the contact action when the formulation `MORTAR_PENALTY` and
```
Let's not do that. AL is simply penalty with an AL problem class.

dschwen commented on modules/contact/doc/content/source/userobjects/PenaltyFrictionUserObject.md:
```diff
@@ -9,17 +9,55 @@ frictional pressure(s) saturates at the Coulomb limit. Note that a high enough
+
+An augmented Lagrange (AL) approach can be used to enforce the contact constraints to a user-prescribed
+tolerance. That tolerance can be the normal gap distance (distance to exact enforcement if in contact) or
+the relative slip tolerance for slipping nodes. The AL approach solves the original MOOSE problem
```
added 

> , in which contact is enforced using a pure penalty approach,

dschwen commented on modules/contact/src/auxkernels/PenaltyMortarUserObjectAux.C:
```diff
@@ -30,58 +36,128 @@ PenaltyMortarUserObjectAux::validParams()
       "user_object",
       "The penalty mortar user object to get values from.  Note that the user object "
       "must implement the corresponding getter function.");
```
Ugh, that is just waiting to be outdated

dschwen replied:
Also `WeightedGapUserObject` is not a registered type

dschwen commented on modules/contact/include/postprocessors/NumAugmentedLagrangeIterations.h:
```diff
@@ -0,0 +1,39 @@
+class AugmentedLagrangianContactProblemInterface;
+
+/**
+ * Get the number of extra augmented lagrange loops. This PP will return zero if the current problem
```
Get the number of extra augmented Lagrange loops around the non-linear solve.

dschwen commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -8,22 +8,61 @@
+      "the contacting boundary. This parameter is defaulted to 'true' in the contact action.");
+  params.addRangeCheckedParam<Real>(
+      "penetration_tolerance",
+      1e-5,
```
Ugh, we cannot make it required, because it is only required for AL. We could sprinkle a bunch of `isParamSetByUser` in there ...

bwspenc replied:
It's ugly, but I think this particular one is important to require the user specifies somehow.

dschwen replied:
Added a check at the action level.

bwspenc replied:
Per our discussion, the thing to do is:
1) Remove the default value here
2) Generate an error if the AL problem is being used and `isParamValid()` is false for this parameter.
3) Add a test for that error, probably just by using cli_args on an existing test.
4) Handle the slip_tolerance parameters in the same way.

dschwen committed:
Address comments (#24168)

dschwen committed:
Address more comments (#24168)

bwspenc commented on modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d-rz/tests:
```diff
@@ -3,7 +3,8 @@
   parallel_scheduling = true
   [thermomechanical_mortar]
-    requirement = 'The system shall be able to solve axisymmetric thermomechanical contact with finite-strain tensor mechanics and heat conduction physics with'
+    requirement = 'The system shall be able to solve axisymmetric thermomechanical contact with '
```
That wasn't a commentary on the current state of the requirement text. Not a major issue. It's just that the phrase 'be able to' doesn't add any value.

recuero replied:
It doesn't add value but it has been enforced during PR reviews quite a bit.

bwspenc replied:
Do you mean that reviewers are asking people to add that?

recuero replied:
Yes.

One could argue that "shall be able to" allows one to skip the conditions under which the "system shall". But, anyway, it seems that, as Daniel showed above, there has been differences of opinion within the CCB.

recuero commented on modules/contact/test/tests/pdass_problems/tests:
```diff
@@ -50,6 +50,13 @@
+    input = 'cylinder_friction_penalty_frictional_al_action.i'
+    expect_err = "Augmented Lagrange parameter was specified, but the selected problem type does not support Augmented Lagrange iterations."
+    cli_args = 'Problem/type=FEProblem'
+    requirement = 'The contact action shall be alert the user if Augmented Lagrage FUUUUUUUUUUU.'
```
FUUUUUUU?

moosebuild commented:
Job [Conda MOOSE (ARM Mac)](https://civet.inl.gov/job/1701766/) on 11e7370 : invalidated by @recuero

moosebuild commented:
Job [Conda MOOSE (ARM Mac)](https://civet.inl.gov/job/1701766/) on 11e7370 : invalidated by @recuero

bwspenc reviewed (changes requested):
Still a few minor things that I think should still be addressed. As we discussed, by far the most important one is getting rid of the default tolerances.

dschwen committed:
Address yet more comments (#24168)

recuero commented on modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C:
```diff
@@ -64,17 +64,10 @@ PenaltyWeightedGapUserObject::PenaltyWeightedGapUserObject(const InputParameters
-  {
-    if (!var.isNodal())
-      paramError(var_name,
-                 "The displacement variables must have degrees of freedom exclusively on "
```
I missed why this piece of code is removed.

dschwen replied:
Because the same check is already present in the parent class.

dschwen replied:
Oh wait.... Did I remove it in the wrong class.. I might have...

dschwen replied:
Fixed. By teh way we've been working hard to move away from `disp_[xyz]` parameters in favor of the `displacements` parameter. Any reason that pattern wasn't used here?

recuero replied:
The only reason is that we repeated old patterns in these objects. So it can be migrated to newer displacements parameter. In fact, that's a task of a contact maintenance issue for the entire module.

moosebuild commented:
Job [Conda MOOSE (ARM Mac)](https://civet.inl.gov/job/1715625/) on c9be92e : invalidated by @recuero

bwspenc reviewed (approved):
Looks good, thanks for addressing my issues.

Follow-on items that we shouldn't forget:
1) Rename those "force" variables to "pressure"
2) Change the default value of "penalty_multiplier" to 100 in the Contact action
